src/compiler/checker.ts

```diff
@@ -29379,98 +29383,22 @@ namespace ts {
                 return typeAsAwaitable.awaitedTypeOfType = type;
             }
 
-            if (type.flags & TypeFlags.Union) {
-                let types: Type[] | undefined;
-                for (const constituentType of (<UnionType>type).types) {
-                    types = append<Type>(types, getAwaitedType(constituentType, errorNode, diagnosticMessage, arg0));
-                }
-
-                if (!types) {
-                    return undefined;
-                }
-
-                return typeAsAwaitable.awaitedTypeOfType = getUnionType(types);
+            const symbol = getGlobalAwaitedSymbol(/*reportErrors*/ false);
+            if (!symbol) {
+                return typeAsAwaitable.awaitedTypeOfType = type;
             }
 
-            const promisedType = getPromisedTypeOfPromise(type);

```

Comment:
I don't agree with this change. This completely removes support for recursively unwrapping a non-native promise, which is a core capability of `await` and our type-system support for it which has existed since the feature was added.

Reply:
Can a non-native promise resolve to another promise? If not, then `await` isn't recursive and I think this change is safe?

If it can, then I concede that implies a [recursive awaited solution](https://github.com/microsoft/TypeScript/pull/35998).

Reply:
Well, non‑native&nbsp;promises can&nbsp;resolve to&nbsp;an&nbsp;object with&nbsp;a&nbsp;`.then` function&nbsp;property, which&nbsp;`await` will&nbsp;resolve&nbsp;recursively.

Reply:
Wouldn't that violate [Promises/A+](https://promisesaplus.com)?

Reply:
@jablko: non-native promise-like implementations aren't all guaranteed to be Promise/A+. In the past, JQuery's `defer` created something promise-like (i.e. it has a callable `then` that accepted fulfill/reject callbacks), but it did *not* recursively unwrap. This is why native `await` and native `Promise` both recursively unwrap (to defend against non-Promise/A+-compliant promise-likes)

Reply:
`await` *is* recursive for non-native promises. In https://tc39.es/ecma262/#await, the value is passed to [`PromiseResolve`](https://tc39.es/ecma262/#sec-promise-resolve) which adopts the foreign promise and performs recursive unwrapping.

Reply:
Thanks! I concede that `await` is recursive for non-compliant promises, however there's only ever been incomplete support for non-compliant promises in TypeScript. TypeScript's native promise type hasn't recursively unwrapped non-compliant promises until [now](https://github.com/microsoft/TypeScript/pull/35998).

In practice, do non-compliant promises justify introducing a new kind of type? For native and A+-compliant promises, `await` isn't recursive because they can't be nested.

Is it more likely that an instance of a non-compliant promise is actually an error? Supporting non-compliant promises can be [error prone](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41960), by masking what would otherwise be errors.